src/ads.jl

```diff
@@ -0,0 +1,554 @@
+                 nsplits::Int, maxsplits::Int, stol::T) where {T, N, M}
+
+Split `node` in half if any element of `x0` has an `adsnorm` greater than `stol`.
+"""
+function decidesplit!(node::ADSTaylorSolution{T, N, M}, dt::T, x0::Vector{TaylorN{T}},
+    nsplits::Int, maxsplits::Int, stol::T) where {T, N, M}
+    # Split criteria for each element of x
+    mask = adsnorm.(x0)
+    # Split
+    if nsplits < maxsplits && any(mask .> stol)
+        halve!(node, dt, x0)
+        # Update number of splits
+        nsplits += 1
+    # No split
+    else
+        leftchild!(node, node.t + dt, node.lo, node.hi, SVector{M, TaylorN{T}}(x0))
+    end
+
+    return nsplits
+end
+
+@inline function set_psol!(::Val{true}, node::ADSTaylorSolution{T, N, M},
+    x::Vector{Taylor1{TaylorN{T}}}) where {T <: Real, N, M}
+    node.p = deepcopy.(x)
+    return nothing
+end
+
+@inline function set_psol!(::Val{false}, node::ADSTaylorSolution{T, N, M},
+    x::Vector{Taylor1{TaylorN{T}}}) where {T <: Real, N, M}
+    node.p = nothing
+    return nothing
+end
+
+function taylorinteg(f!, q0::ADSTaylorSolution{T, N, M}, t0::T, tmax::T, order::Int,
+    stol::T, abstol::T, params = nothing; maxsplits::Int = 10, maxsteps::Int = 500,
+    parse_eqs::Bool = true, dense::Bool = true) where {T <: Real, N, M}
+
+    # Initialize the vector of Taylor1 expansions
+    t = t0 + Taylor1( T, order )
+    x = Array{Taylor1{TaylorN{T}}}(undef, M)
+    dx = Array{Taylor1{TaylorN{T}}}(undef, M)
+    @inbounds for i in eachindex(x)
+        x[i] = Taylor1( q0.x[i], order )
+        dx[i] = Taylor1( zero(q0.x[i]), order )
+    end
+
+    # Determine if specialized jetcoeffs! method exists
+    parse_eqs, rv = _determine_parsing!(parse_eqs, f!, t, x, dx, params)
+
+    return _taylorinteg!(Val(dense), f!, q0, t0, tmax, order, stol, abstol, rv,
+        params; parse_eqs, maxsplits, maxsteps)
+end
+
+function _taylorinteg!(dense::Val{D}, f!, q0::ADSTaylorSolution{T, N, M}, t0::T,
+    tmax::T, order::Int, stol::T, abstol::T, rv::RetAlloc{Taylor1{TaylorN{T}}},
+    params = nothing; parse_eqs::Bool = true, maxsplits::Int = 10,
+    maxsteps::Int = 500) where {T <: Real, D, N, M}
+
+    # Allocation
+    δt = Vector{T}(undef, maxsplits)
+    t = Vector{Taylor1{T}}(undef, maxsplits)
+    x0 = Matrix{TaylorN{T}}(undef, M, maxsplits)
+    x = Matrix{Taylor1{TaylorN{T}}}(undef, M, maxsplits)
+    dx = Matrix{Taylor1{TaylorN{T}}}(undef, M, maxsplits)
+    xaux = Matrix{Taylor1{TaylorN{T}}}(undef, M, maxsplits)
+    @inbounds for j in eachindex(t)
+        δt[j] = zero(T)
+        t[j] = t0 + Taylor1( T, order )
+        @inbounds for i in axes(x0, 1)
+            x0[i, j] = q0.x[i]
+            x[i, j] = Taylor1( q0.x[i], order )
+            dx[i, j] = Taylor1( zero(q0.x[i]), order )
+            xaux[i, j] = Taylor1( zero(q0.x[i]), order )
+        end
+    end
+    # IMPORTANT: each split needs its own RetAlloc
+    rvs = [deepcopy(rv) for _ in 1:maxsplits]
```

Comment:
Is it clear why in the end a separate `RetAlloc` is needed for each split? If it's just some aliasing occurring in the arguments of `jetcoeffs!` maybe there's a way to work around it. Or maybe it's something else; either way I think it can help to understand where this is coming from.

Reply:
My naive guess is that you want some sort of "thread safety" in the tree context, which I'm unable to formulate properly...

Reply:
My guess is that it is not a "thread safety" issue since splits are processed sequentially. When I started doing ADS integrations in `NEOs` I noticed unexpected results that were solved by introducing one `RetAlloc` per split. So, there is something in `NEOs` dynamical functions that tests here do not notice and affect the `RetAlloc` object.